### PR TITLE
[AArch64] Add updated FEAT_SVE_B16B16 and begin replacement of 'b16b16' flag

### PR DESCRIPTION
**Pull request by SpencerAbson**
This patch adds FeatureSVEB16B16 to the AArch64 backend in order to represent the new behavior of FEAT_SVE_B16B16 (as described in the latest [Armv9.4 extensions documentation](https://developer.arm.com/documentation/109697/0100/Feature-descriptions/The-Armv9-4-architecture-extension?lang=en#md461-the-armv94-architecture-extension__FEAT_SVE_B16B16)) as well as a 'sve-b16b16' flag to enable it. 

The predication of non-widening SVE BFloat16 instructions has changed to require this feature, instead of the previously required and soon-to-be-removed FeatureB16B16 which is enabled by the 'b16b16' flag. Therefore, this change weakens the 'b16b16' flag in favour of 'sve-b16b16'. Existing tests that are effected by this have been modified to use and/or expect 'sve-b16b16', and new tests have been added to verify the behavior and implementation of 'sve-b16b16'. 

This patch is in response to the response to the following changes.

The architecture features previously enabled by FEAT_SVE_B16B16 have been relaxed such that it now implements:
      - With FEAT_SVE2 : SVE non-widening BFloat16 instructions in Non-streaming SVE mode
      - With FEAT_SME2: SVE non-widening BFloat16 instructions when the PE is in Streaming SVE mode and SME
        Z-targeting multi-vector non-widening BFloat16 instructions.
      - **It no longer implements** SME ZA-targeting non-widening BFloat16 instructions.      

The SME ZA-targeting non-widening BFloat16 instructions are implemented by the new FEAT_SME_B16B16, **this patch does not change how this architecture feature is enabled** ('+b16b16+sme2'). Only those that are implemented by FEAT_SVE_B16B16 have been changed to require 'sve-b16b16' instead of 'b16b16'.

New flags must be created to represent FEAT_SVE_B16B16 and FEAT_SME_B16B16:
      - 'sve-b16b16' enables the updated FEAT_SVE_B16B16 (described here)
      - 'sme-b16b16' will enable the new FEAT_SME_B16B16
      - **This patch includes 'sve-b16b16' only**
   
A future patch will add 'sme-b16b16', SME ZA-targeting non-widening BFloat16 instructions would then be guarded by '+sme-b16b16+sme2', and 'b16b16' can be removed.

### CODE REVIEW

File: clang/include/clang/Basic/arm_sve.td [hunk 1]
```diff
@@ -2172,7 +2172,7 @@ let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2" in {
   def SVFCLAMP_X4 : SInst<"svclamp[_single_{d}_x4]",  "44dd",   "hfd",      MergeNone, "aarch64_sve_fclamp_single_x4",  [IsStreaming], []>;
 }
 
-let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2,b16b16"in {
+let SVETargetGuard = InvalidMode, SMETargetGuard = "sme2,sve-b16b16"in {
```
Comment (CarolineConcatto) [inline clang/include/clang/Basic/arm_sve.td:2175]: It is missing the multi vector targeting Z. For some reason they are not under b16b16.
But they were added in this commit f7392f40f3f6d5d4f.
I can see them being changes in LLVM in this patch
// SME2 Z-targeting non-widening BFloat16 instructions
let Predicates = [HasSME2, HasSVEB16B16] in {

**Review by CarolineConcatto — COMMENTED** [review]
Thank you Spencer for the patch.
I believe we missed the  multi vector instructions targeting BF16.

File: clang/include/clang/Basic/arm_sve.td (hunk 1 above)
Comment (SpencerAbson) [reply clang/include/clang/Basic/arm_sve.td:2175]: Thanks for pointing this out.

File: clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// RUN: %clang_cc1 -triple aarch64-none-linux-gnu -target-feature +sve -target-feature +sve2 -verify -verify-ignore-unexpected=error,note -emit-llvm -o - %s
```
[the remark this answers is not in the document]
Comment (SpencerAbson) [reply clang/test/Sema/aarch64-sve2p1-intrinsics/acle_sve2p1_b16b16.cpp:1]: Will do :+1:

File: clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfclamp.c [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfclamp.c:7]: Not related to you patch, but the instructions that you changed  now only require sve2.
I believe you can change the run line to be only sve2.
The only ones that are sve2.1 are in section
SVE2.1 instruction intrinsics in the ACLE https://github.com/ARM-software/acle/blob/main/main/acle.md

File: clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmla_lane.c [hunk 4]
```diff
@@ -1,10 +1,10 @@
...
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmla_lane.c:7]: Same here.
s/sve2p1/sve2/
Please!

File: clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmls_lane.c [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s
+// RUN: %clang_cc1 -DSVE_OVERLOADED_FORMS -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -emit-llvm -o - -x c++ %s | opt -S -p mem2reg,instcombine,tailcallelim | FileCheck %s -check-prefix=CPP-CHECK
+// RUN: %clang_cc1 -triple aarch64 -target-feature +sve -target-feature +sve2 -target-feature +sve2p1 -target-feature +sve-b16b16 -disable-O0-optnone -Werror -Wall -o /dev/null %s
```
Comment (CarolineConcatto) [inline clang/test/CodeGen/aarch64-sve2p1-intrinsics/acle_sve2p1_bfmls_lane.c:7]: Same here.
s/sve2p1/sve2/
Please!

File: clang/test/Sema/aarch64-sve2-intrinsics/acle_sve2_bfloat.cpp [hunk 6]
```diff
@@ -26,4 +26,4 @@ void test_bfloat(const bfloat16_t *const_bf16_ptr, svbfloat16_t bf16, svbfloat16
   // expected-error@+2 {{'svwhilewr_bf16' needs target feature (sve2,bf16)|(sme,bf16)}}
   // overload-error@+1 {{'svwhilewr' needs target feature (sve2,bf16)|(sme,bf16)}}
   SVE_ACLE_FUNC(svwhilewr,_bf16,,)(const_bf16_ptr, const_bf16_ptr);
-}
+}
```
Comment (CarolineConcatto) [inline clang/test/Sema/aarch64-sve2-intrinsics/acle_sve2_bfloat.cpp:29]: Unrelated change

**Review by CarolineConcatto — APPROVED** [review]
Thank you Spencer for the work.
I left some comments about the sve2.1 feature flags being used for the sve2. But I dont think you need to change in this patch. Because it is unrelated
Maybe you create another patch a NFC to move all the sve2p1 tests and run lines to be only SVE2?
I believe when we changed the feature flag we forgot to move the tests and also to fix all run lines.

**Comment by vfdff** [comment]
hi, SpencerAbson
    How to enable the behavior of **FEAT_SVE_B16B16** after this commit  ? I try **clang -march=armv8.5a+sve2+sve-b16b16**, but it seems doesn't take effect. https://gcc.godbolt.org/z/sadrshqvM 
  > Or the only method to effect is **-march=armv8.5a+sve2+bf16 -O3 -Xclang -target-feature -Xclang +sve-b16b16** ?

**Comment by SpencerAbson** [comment]
Hi @vfdff

You were correct the first time, `-march=armv8.5a+sve2+sve-b16b16` will enable the instructions (and intrinsics) under [FEAT_SVE_B16B16](https://developer.arm.com/documentation/109697/0100/Feature-descriptions/The-Armv9-4-architecture-extension?lang=en#md461-the-armv94-architecture-extension__FEAT_SVE_B16B16).

However, please refer to the [ACLE ](https://github.com/ARM-software/acle/blob/latest-release/main/acle.md) concerning the use of the `__bf16` type. 
> The __bf16 type is only available when the __ARM_BF16_FORMAT_ALTERNATIVE feature macro is defined. When it is available it can only be used by the ACLE intrinsics ; it cannot be used with standard C operators. It is expected that arithmetic using standard C operators be used using a single-precision floating point format and the value be converted to __bf16 when required using ACLE intrinsics. 

Please bare in mind that the B16B16 support is in alpha state, and may be changed in the future.  If you would like more information about Arm intrinsics support for Bfloat16, feel free to reach out to me personally (spencer.abson@arm.com). It might also be useful to have a look at recent changes to the [main branch of the ACLE](https://github.com/ARM-software/acle/blob/main/main/acle.md) to get an idea about how the interface should look in the near future.

Thanks
